File: apps/mobile/lib/__tests__/dropCount.test.ts
import { contributesVolume, countsAsSet, type LoggedSet } from '../sessions';

const s = (over: Partial<LoggedSet> = {}): LoggedSet => ({
  exercise_id: 'bench',
  position: 0,
  set_type: 'working',
  reps: 5,
  weight_kg: 100,
  seconds: null,
  distance_m: null,
  rir: null,
  rpe: null,
  notes: '',
  completed: true,
  ...over,
});

describe('a drop is part of the set above it', () => {
  it('does not count as a set, but does contribute volume', () => {
    // The pair IS the change. Asserting only the first half would pass against
    // a version that also discarded the drop's work — which is the mistake the
    // split exists to prevent, and the one that would lose real training.
    const drop = s({ set_type: 'drop' });
    expect(countsAsSet(drop)).toBe(false);
    expect(contributesVolume(drop)).toBe(true);
  });

  it('leaves every other performed set alone', () => {
    for (const t of ['working', 'backoff', 'amrap', 'failure'] as const) {
      expect(countsAsSet(s({ set_type: t }))).toBe(true);
      expect(contributesVolume(s({ set_type: t }))).toBe(true);
    }
  });

  it('excludes a warm-up from both', () => {
    const warm = s({ set_type: 'warmup' });
    expect(countsAsSet(warm)).toBe(false);
    expect(contributesVolume(warm)).toBe(false);
  });

  it('excludes anything not performed from both', () => {
    // Planned but not ticked. A template opens with every set false.
    const planned = s({ completed: false });
    expect(countsAsSet(planned)).toBe(false);
    expect(contributesVolume(planned)).toBe(false);
  });
});

File: backend/internal/modules/session/dropcount_test.go
package session

import "testing"

func dp(v int) *int         { return &v }
func dw(v float64) *float64 { return &v }

// A drop is part of the set it came off — one approach to the bar, one rest
// period — so it does not add to the number the athlete counts. Its work still
// does: the weight was moved.
//
// Both halves in one test on purpose. Asserting only the count would pass
// against a change that excluded drops from the volume too, which is the
// mistake this split exists to prevent.
func TestADropIsNotASetButItsWorkCounts(t *testing.T) {
	sets := []Set{
		{ExerciseID: "bench", SetType: SetTypeWorking, Reps: dp(3), WeightKg: dw(100), Completed: true},
		{ExerciseID: "bench", SetType: SetTypeDrop, Reps: dp(8), WeightKg: dw(80), Completed: true},
		{ExerciseID: "bench", SetType: SetTypeWorking, Reps: dp(3), WeightKg: dw(100), Completed: true},
	}
	v := Summarise(sets)
	if v.WorkingSets != 2 {
		t.Fatalf("counted %d sets, want 2 — a drop is part of the set above it", v.WorkingSets)
	}
	// 3x100 + 8x80 + 3x100 = 1240. The drop's 640 is in there.
	if v.TonnageKg != 1240 {
		t.Fatalf("tonnage %v, want 1240 — the drop's work must still count", v.TonnageKg)
	}
	if v.TotalReps != 14 {
		t.Fatalf("reps %d, want 14 — the drop's reps were performed", v.TotalReps)
	}
}

// Every other set type still counts. The exclusion is drops specifically, not
// "anything unusual".
func TestOnlyDropsAreExcludedFromTheCount(t *testing.T) {
	for _, st := range []SetType{SetTypeWorking, SetTypeBackoff, SetTypeAMRAP, SetTypeFailure} {
		v := Summarise([]Set{{ExerciseID: "b", SetType: st, Reps: dp(5), WeightKg: dw(60), Completed: true}})
		if v.WorkingSets != 1 {
			t.Fatalf("%s counted %d, want 1", st, v.WorkingSets)
		}
	}
	// And a warm-up still contributes nothing at all.
	v := Summarise([]Set{{ExerciseID: "b", SetType: SetTypeWarmup, Reps: dp(10), WeightKg: dw(20), Completed: true}})
	if v.WorkingSets != 0 || v.TonnageKg != 0 {
		t.Fatalf("a warm-up contributed sets=%d tonnage=%v", v.WorkingSets, v.TonnageKg)
	}
}
